### PR TITLE
Add filter page

### PR DESCRIPTION
**Pull request by AdriB02**
# Filter Page feature and update Swipe Page
## Description
- What has been changed?
> ### New Model and Integration:
>  * `FilterPage.kt`: Add a new FilterPage to select filters for the recipes dispalyed in the swipe page. Currently, we can only apply one filter per category due to MealDB's limitations, and all filters do not affect the recipes at this time, as we need to wait for our database to be completed before making updates.
image

> * `SwipePage.kt`: This PR introduces InputChips to showcase selected filters (e.g., category, time range, price range, and difficulty) in a single, organized row. When multiple filters are applied, the row scrolls horizontally, ensuring all filters remain accessible on any screen size.
 Each InputChip displays the current filter and includes a close icon, allowing users to quickly remove filters as needed. Removing a filter updates recipesViewModel in real time, providing an intuitive and responsive user experience for managing filters within the UI.

> image

> * [`Filter.kt`](diffhunk://#diff-49b50884afbc98e727fddc88d73a10133be7920b64f8f79d4039920e58a40baaR1-R50): Introduced a new `Filter` class to store filter types and ranges, along with a `FloatRange` class and `Difficulty` enum.
> * [`RecipesViewModel.kt`](diffhunk://#diff-43b7e4735c753181da60b2909d6610988ff0c96c3e73a8830233fb7e6573a277R3-R8): Integrated the new `Filter` model into `RecipesViewModel`, adding state flows for filter and categories.
> * [`MainActivity.kt`](diffhunk://#diff-0859fc4e472afc96e90cddd4bc16c22a149a751d8794a52a66c5053ec043660dR23): Added navigation to the `FilterPage` in the main application. 

> ### Minor Fixes:
> * [`MealDBRecipesRepository.kt`](diffhunk://#diff-f560bede89f5f0cc2e8313fa7168dae0fc1830d032dfd72f48f85d391b08cc4cR6): Corrected the URL and JSON key for fetching categories from the MealDB API.

> ### Tests
> This pull request introduces new test cases for the `FilterPage` and `SwipePage` components, adds a new `Filter` model, and integrates the `FilterPage` into the main application. The changes are primarily focused on enhancing the filtering functionality and ensuring proper test coverage.

> * [`FilterPageTest.kt`](diffhunk://#diff-34d286c816202362fa625f3240cecb98aff48911d270181adae300d05bd9467eR1-R286): Added comprehensive test cases for `FilterPage`, including tests for UI elements and ViewModel updates.
> * [`SwipePageTest.kt`](diffhunk://#diff-8a10ba4b62df25b6bc5b66cb37c782c9d508feeccba9a68491a3653a87d7efc2R4-R14): Added tests for displaying and hiding filter chips, and initialized filter values in setup. 

- Why was this change made?
> This change enhances the user experience on the Swipe Page by allowing more flexible and customizable filtering options. With the new FilterPage and InputChips display, users can now quickly apply and view filters directly within the swipe view, improving accessibility and control over displayed recipes. The introduction of InputChips allows easy filter management with a clean, organized UI, and the horizontal scroll ensures visibility on various screen sizes. Integrating filters into the ViewModel also lays the groundwork for future feature scalability and testing.

## Checklist
- [x] Added comprehensive unit and UI tests for new features and integrations.
- [x] Updated documentation to cover new components and functionality.
- [x] Ensured all CI checks pass with the added code.
- [x] Linked issue/feature: #123

### CODE REVIEW

File: app/src/main/java/com/android/sample/ui/filter/FilterPage.kt
Comment (Ylanv) [inline app/src/main/java/com/android/sample/ui/filter/FilterPage.kt:1]: Overall, I really like the range feature for Time and Price, it’s a great addition! One enhancement we could consider for the future is allowing users to click on the number at either end of the range to input a value directly with their keyboard, giving them more precision.

image

For the checkboxes, limiting selection to one option feels a bit restrictive. Users might want to view multiple options, like both 'easy' and 'medium' recipes. It could also improve the user experience if double-clicking a selected checkbox would deselect it (as shown in the video). Additionally, the categories section feels slightly unstructured without a clear order, which might affect usability. Finally, adding a 'Save' button could make the interface more intuitive for users, giving them a clear way to confirm their selections. This might help streamline the experience and make the interaction feel more complete.

These are just some ideas to consider as we continue refining, overall, great work!

https://github.com/user-attachments/assets/fecd611d-2f14-4b1a-aad1-8900a8415569

File: app/src/main/java/com/android/sample/ui/filter/FilterPage.kt [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+@Composable
+fun FilterBox(paddingValues: PaddingValues, recipesViewModel: RecipesViewModel) {
+  val filter by recipesViewModel.filter.collectAsState()
+  Column(
+      modifier =
+          Modifier.padding(paddingValues).fillMaxSize().verticalScroll(rememberScrollState())) {
+        ValueRangeSlider(
+            modifier = Modifier.testTag("timeRangeSlider"),
+            name = TIME_RANGE_NAME,
+            min = TIME_RANGE_MIN,
```
Comment (Ylanv) [inline app/src/main/java/com/android/sample/ui/filter/FilterPage.kt:110]: The **minimum** and **maximum** values for the Time and Price ranges are currently set as fixed values. It might be beneficial to conduct a bit of research on typical minimum and maximum price/duration values in our database beforehand to ensure these ranges are as coherent and relevant as possible. This could enhance the feature's usability and accuracy for users.

File: app/src/main/java/com/android/sample/ui/filter/FilterPage.kt [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+
+    Spacer(modifier = Modifier.height(8.dp))
+
+    // Range slider with minimum 0 and maximum 120 minutes
+    RangeSlider(
+        modifier = modifier.fillMaxWidth(),
+        value = sliderPosition,
+        onValueChange = {
+          sliderPosition = it
+          updateRange(it.start, it.endInclusive)
```
Comment (Ylanv) [inline app/src/main/java/com/android/sample/ui/filter/FilterPage.kt:183]: From my understanding, each time the user adjusts the slider, this triggers `recipesViewModel.updatePriceRange(newMin, newMax)`, which currently just updates `_filter = MutableStateFlow(Filter())`. However, if updating this value eventually triggers a new fetch, we could end up with a significant number of unnecessary API calls. Since from my understanting this part is not implemented we need to be carefull.

For optimization, it might be more efficient to introduce a 'Save' button that applies all selected filters at once, reducing the API requests and improving performance.

File: app/src/main/java/com/android/sample/ui/filter/FilterPage.kt [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+  }
+  val orderedDifficulties = listOf(Difficulty.Easy, Difficulty.Medium, Difficulty.Hard)
+
+  if (recipesViewModel.filter.value.difficulty != Difficulty.Undefined) {
+    difficultyLevels[recipesViewModel.filter.value.difficulty] = true
+  }
+
+  Column(modifier = Modifier.padding(16.dp), horizontalAlignment = Alignment.CenterHorizontally) {
+    Text(
+        text = "Difficulty",
```
Comment (Ylanv) [inline app/src/main/java/com/android/sample/ui/filter/FilterPage.kt:217]: Great job handling the magic strings in most areas, it looks like just this one was missed. Easy to overlook! Updating it for consistency would keep everything clean and maintainable.

File: app/src/main/java/com/android/sample/ui/filter/FilterPage.kt [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+        style = MaterialTheme.typography.titleMedium,
+        modifier = Modifier.align(Alignment.Start))
+
+    // Display each checkbox with the corresponding difficulty name
+    FlowRow(
+        modifier = Modifier.fillMaxWidth(),
+        horizontalArrangement = Arrangement.Center,
+        verticalArrangement = Arrangement.Center,
+        maxItemsInEachRow = 3) {
```
Comment (Ylanv) [inline app/src/main/java/com/android/sample/ui/filter/FilterPage.kt:226]: magic number

File: app/src/main/java/com/android/sample/ui/filter/FilterPage.kt [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+              Spacer(modifier = Modifier.height(4.dp))
+
+              Text(
+                  text = difficulty.toString(),
+                  style = MaterialTheme.typography.bodyMedium,
+                  color = MaterialTheme.colorScheme.onPrimary) // Display the difficulty name
+              Checkbox(
+                  modifier = Modifier.testTag("difficultyCheckbox${difficulty}"),
+                  checked = difficultyLevels[difficulty]!!,
+                  onCheckedChange = { isChecked ->
```
Comment (Ylanv) [inline app/src/main/java/com/android/sample/ui/filter/FilterPage.kt:240]: The `CheckboxDifficulty` and `CheckboxCategories` methods have a lot of overlapping code, particularly in how the checkboxes and categories are displayed. It might be worth modularizing this shared logic into a single reusable component. This would help streamline updates, improve readability, and maintain consistency across both methods. Great work overall, just a small suggestion to enhance maintainability!"

**Review by Ylanv — COMMENTED** [review]
Great job on this extensive PR, it’s clear that a lot of effort went into it! I’ve added a few questions and suggestions for consideration. Please feel free to share your thoughts or disagree in the comments. Excellent work overall!

File: app/src/main/res/drawable/filter.png
[the remark this answers is not in the document]
Comment (AdriB02) [reply app/src/main/res/drawable/filter.png:1]: I agree, I will update it now !

File: app/src/main/java/com/android/sample/model/filter/Filter.kt [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+data class FloatRange(var min: Float, var max: Float, var minBorn: Float, var maxBorn: Float) {
+  init {
+    require(min <= max) { MIN_SHOULD_NOT_BE_GREATER_THAN_MAX }
+  }
+
+  fun update(newMin: Float, newMax: Float) {
+    require(0 <= newMin) { NEW_MIN_SHOULD_NOT_BE_NEGATIVE }
+    require(0 <= newMax) { NEW_MAX_SHOULD_NOT_BE_NEGATIVE }
+    require(newMin <= newMax) { NEW_MIN_SHOULD_NOT_EXCEED_MAX }
+    if (minBorn == -1f && maxBorn == -1f) {
```
[the remark this answers is not in the document]
Comment (AdriB02) [reply app/src/main/java/com/android/sample/model/filter/Filter.kt:27]: Done !

File: app/src/main/java/com/android/sample/ui/filter/FilterPage.kt (hunk 1 above)
Comment (AdriB02) [reply app/src/main/java/com/android/sample/ui/filter/FilterPage.kt:110]: Thank you for your feedback! The current minimum and maximum values for the Time and Price ranges are indeed set as temporary fixed values. I plan to update them later once the backend management system is in place.

File: app/src/main/java/com/android/sample/ui/filter/FilterPage.kt (hunk 2 above)
Comment (AdriB02) [reply app/src/main/java/com/android/sample/ui/filter/FilterPage.kt:183]: Thank you for the detailed feedback! You’re correct that, as it stands, each slider adjustment updates ? 
`recipesViewModel.updatePriceRange(newMin, newMax)`. Once the backend is available, I’ll optimize this within `RecipeViewModel` to minimize API calls by using a range of values that only triggers updates when necessary.

Regarding the 'Save' button, I actually prefer not to include it, as the main goal of my UI design is to keep interactions as simple as possible. After reviewing similar applications, it seems that saving changes automatically (rather than using a button) creates a smoother user experience, especially since users can directly delete filters from the Swipe Page. For now, I’ll leave the current setup as is, and when the backend is complete, I’ll handle filtering in a way that’s dependent on the database.

File: app/src/main/java/com/android/sample/ui/filter/FilterPage.kt (hunk 4 above)
Comment (AdriB02) [reply app/src/main/java/com/android/sample/ui/filter/FilterPage.kt:226]: Fixed !

File: app/src/main/java/com/android/sample/ui/filter/FilterPage.kt (hunk 5 above)
Comment (AdriB02) [reply app/src/main/java/com/android/sample/ui/filter/FilterPage.kt:240]: It was a bit challenging, but I managed to complete it!

File: app/src/androidTest/java/com/android/sample/screen/SwipePageTest.kt [hunk 7]
```diff
@@ -79,6 +83,15 @@ class SwipePageTest : TestCase() {
 
     `when`(mockNavigationActions.currentRoute()).thenReturn(Route.SWIPE)
     `when`(mockNavigationActions.navigateTo(Route.AUTH)).then {}
+
+    // Init the filter
+    recipesViewModel.updateTimeRange(0f, 100f)
+    recipesViewModel.updateTimeRange(5f, 99f)
+    recipesViewModel.updatePriceRange(0f, 100f)
+    recipesViewModel.updatePriceRange(5f, 52f)
```
Comment (akshire) [inline app/src/androidTest/java/com/android/sample/screen/SwipePageTest.kt:91]: These constants could go in the file for the constants

File: app/src/androidTest/java/com/android/sample/screen/SwipePageTest.kt [hunk 8]
```diff
@@ -79,6 +83,15 @@ class SwipePageTest : TestCase() {
...
+    recipesViewModel.updateDifficulty(Difficulty.Easy)
+    recipesViewModel.updateCategory("Dessert")
```
Comment (akshire) [inline app/src/androidTest/java/com/android/sample/screen/SwipePageTest.kt:93]: Also these constants could go in the file for constants

File: app/src/main/java/com/android/sample/model/recipe/MealDBRecipesRepository.kt [hunk 9]
```diff
@@ -78,7 +79,7 @@ class MealDBRecipesRepository(private val client: OkHttpClient) : RecipesReposit
    */
   private fun parseMealDBJsonToCategory(json: JSONObject): List<String> {
     val parsedListOfCategories = mutableListOf<String>()
-    val listOfCategories = json.getJSONArray(MEAL_DB_ARRAY_NAME)
+    val listOfCategories = json.getJSONArray(MEAL_DB_CATEGORY_ARRAY)
```
Comment (akshire) [inline app/src/main/java/com/android/sample/model/recipe/MealDBRecipesRepository.kt:82]: Great catch:) !!

File: app/src/main/java/com/android/sample/model/filter/Filter.kt
[the remark this answers is not in the document]
Comment (AdriB02) [reply app/src/main/java/com/android/sample/model/filter/Filter.kt:1]: I use FloatRange because the rangeSlide requires float values. I prefer casting when displaying values instead of modifying them. IntRange could also be an option ^^.

File: app/src/test/java/com/android/sample/model/filter/FilterTest.kt
[the remark this answers is not in the document]
Comment (AdriB02) [reply app/src/test/java/com/android/sample/model/filter/FilterTest.kt:1]: Yes right, I will add these require !

File: app/src/main/java/com/android/sample/ui/filter/FilterPage.kt
Comment (AdriB02) [reply app/src/main/java/com/android/sample/ui/filter/FilterPage.kt:1]: Implementing a keyboard input range is a great idea and can be added as P2 for a future sprint. Regarding the checkboxes, as mentioned in the PR description, we can only select one filter per category due to the limitations of MealDB. This will be addressed once our database is completed.

File: app/src/main/java/com/android/sample/ui/swipePage/SwipePage.kt [hunk 10]
```diff
@@ -160,6 +161,163 @@ fun RecipeDisplay(
...
+                    modifier = Modifier.testTag("priceRangeChip"),
+                    selected = false,
+                    onClick = {},
+                    label = {
+                      Text(
+                          text =
+                              "${filter.priceRange.min.toInt()} - ${filter.priceRange.max.toInt()} $",
+                          style = MaterialTheme.typography.bodySmall,
+                          color = MaterialTheme.colorScheme.onSecondary,
+                      )
```
Comment (akshire) [inline app/src/main/java/com/android/sample/ui/swipePage/SwipePage.kt:240]: Couldn't this be extracted in a function, this would make the Page more concise and more readable.

File: app/src/main/java/com/android/sample/ui/swipePage/SwipePage.kt [hunk 11]
```diff
@@ -160,6 +161,163 @@ fun RecipeDisplay(
...
+                InputChip(
+                    modifier = Modifier.testTag("difficultyChip"),
+                    selected = false,
+                    onClick = {},
+                    label = {
+                      Text(
+                          text = "${filter.difficulty}",
+                          style = MaterialTheme.typography.bodySmall,
+                          color = MaterialTheme.colorScheme.onSecondary,
+                      )
```
Comment (akshire) [inline app/src/main/java/com/android/sample/ui/swipePage/SwipePage.kt:272]: c.f. Extraction comment from same file

File: app/src/main/java/com/android/sample/ui/swipePage/SwipePage.kt [hunk 12]
```diff
@@ -160,6 +161,163 @@ fun RecipeDisplay(
...
+                InputChip(
+                    modifier = Modifier.testTag("categoryChip"),
+                    selected = false,
+                    onClick = {},
+                    label = {
+                      Text(
+                          text = "${filter.category}",
+                          style = MaterialTheme.typography.bodySmall,
+                          color = MaterialTheme.colorScheme.onSecondary,
+                      )
```
Comment (akshire) [inline app/src/main/java/com/android/sample/ui/swipePage/SwipePage.kt:302]: c.f. extraction comment on same file

File: app/src/main/java/com/android/sample/MainActivity.kt [hunk 13]
```diff
@@ -53,7 +54,7 @@ fun PlateSwipeApp() {
 
   val userViewModel = UserViewModel.Factory.create(UserViewModel::class.java)
 
-  NavHost(navController = navController, startDestination = Route.AUTH) {
+  NavHost(navController = navController, startDestination = Route.SWIPE) {
```
Comment (akshire) [inline app/src/main/java/com/android/sample/MainActivity.kt:57]: Correct me if I'm wrong but shouldn't this line be:
```suggestion
 `NavHost(navController = navController, startDestination = Route.AUTH) {
```
 ?

File: app/src/main/java/com/android/sample/model/recipe/MealDBRecipesRepository.kt [hunk 14]
```diff
@@ -190,7 +191,7 @@ class MealDBRecipesRepository(private val client: OkHttpClient) : RecipesReposit
   }
 
   override fun listCategories(onSuccess: (List<String>) -> Unit, onFailure: (Exception) -> Unit) {
-    val url = "$MEAL_DB_URL/categories.php"
+    val url = "https://www.themealdb.com/api/json/v1/1/categories.php"
```
Comment (akshire) [inline app/src/main/java/com/android/sample/model/recipe/MealDBRecipesRepository.kt:194]: This can be transformed to this :
```suggestion
   val url = MEAL_DB_URL +"categories.php"
```

File: app/src/test/java/com/android/sample/model/recipe/MealDBRecipesRepositoryTest.kt [hunk 15]
```diff
@@ -523,7 +523,7 @@ class MealDBRecipesRepositoryTest {
     `when`(mockResponseBody.string())
         .thenReturn(
             """{
-                "meals": [
+                "categories": [
```
Comment (akshire) [inline app/src/test/java/com/android/sample/model/recipe/MealDBRecipesRepositoryTest.kt:526]: Again Great catch :) !!

**Review by akshire — CHANGES_REQUESTED** [review]
### Overall Review
This is a well-crafted PR with thoughtful attention to detail. The recipe filtering feature is notably polished, and the interface presents a significant enhancement over our initial Figma design. It’s clear that you’ve put effort into making the filtering both functional and user-friendly. This work will be a valuable addition to the app, providing users with a seamless way to browse recipes by category.

### Looking Forward to Firestore Integration
I’m excited to see how this feature will integrate with the Firestore-based recipe database that I’m currently developing. With Firestore’s dynamic data handling and real-time update capabilities, users will experience even more robust filtering across a diverse range of categories. This integration could open up future possibilities, like personalized recipe recommendations and data-driven insights, making the app a more powerful and interactive tool for users.

### Feedback on the Cost Filter Bar
I appreciate the precision you’ve achieved with the cost filter bar, especially by using float values instead of integers, as @Ylanv  pointed out :). This level of accuracy is commendable, but I believe there are some broader considerations for how we handle cost filtering. Here’s my feedback on why a symbolic price range approach might serve our goals better:


**User Experience and Simplicity**: For many users, entering an exact price for a recipe may feel like an unnecessary complication. Using symbols to represent general price ranges (such as $, $$, $$$) would simplify the process for users, especially for those who create recipes. They wouldn’t need to estimate a precise cost, which can vary widely depending on factors like ingredient availability and quantity. A range-based system would give users a straightforward way to classify their recipes, enhancing usability and accessibility.

**Scalability Across Regions**: Given that we plan to expand the app’s reach to multiple countries, a symbolic price range would provide consistency across regions. Absolute prices are likely to vary significantly from one location to another due to local economic factors. However, relative price categories tend to remain stable—what’s perceived as "expensive" or "budget-friendly" in one country is generally seen similarly in others. For instance, an “expensive” recipe in Switzerland is likely to be categorized as “expensive” in Italy too, even if the precise cost differs. This would make our app adaptable without requiring constant adjustments to account for currency or cost-of-living differences.

In summary, while the current bar offers precision, I suggest that adopting a symbol-based range system would better align with our goals of creating a user-friendly, globally adaptable app. This approach would allow users to classify their recipes easily while offering a consistent experience across different regions. Thank you for your great work on this PR—I’m looking forward to seeing the continued evolution of this feature!

File: app/src/androidTest/java/com/android/sample/screen/SwipePageTest.kt (hunk 7 above)
Comment (akshire) [reply app/src/androidTest/java/com/android/sample/screen/SwipePageTest.kt:91]: Oups my bad I didn't see that it was in the test file :)

File: app/src/main/java/com/android/sample/model/recipe/MealDBRecipesRepository.kt (hunk 14 above)
Comment (AdriB02) [reply app/src/main/java/com/android/sample/model/recipe/MealDBRecipesRepository.kt:194]: You're right; I changed it for testing and forgot to reset it to the original.

File: app/src/main/java/com/android/sample/MainActivity.kt (hunk 13 above)
Comment (AdriB02) [reply app/src/main/java/com/android/sample/MainActivity.kt:57]: Yes, I modified it for testing and forgot to switch it back.

File: app/src/main/java/com/android/sample/ui/swipePage/SwipePage.kt (hunk 10 above)
Comment (AdriB02) [reply app/src/main/java/com/android/sample/ui/swipePage/SwipePage.kt:240]: Done! And great idea, I’ve extracted it into a function to make the page cleaner and more readable.

File: app/src/main/java/com/android/sample/ui/swipePage/SwipePage.kt (hunk 11 above)
Comment (AdriB02) [reply app/src/main/java/com/android/sample/ui/swipePage/SwipePage.kt:272]: Done !

File: app/src/main/java/com/android/sample/ui/swipePage/SwipePage.kt (hunk 12 above)
Comment (AdriB02) [reply app/src/main/java/com/android/sample/ui/swipePage/SwipePage.kt:302]: Done !

File: app/src/androidTest/java/com/android/sample/screen/SwipePageTest.kt (hunk 7 above)
Comment (AdriB02) [reply app/src/androidTest/java/com/android/sample/screen/SwipePageTest.kt:91]: I see what you mean but for testing, I think it's okay keeping them here.

**Comment by AdriB02** [comment]
Thanks for your review! I appreciate your detailed feedback, and I agree that using symbols like $, $$, $$$ is a great idea for simplifying the cost filter. For now, I’ll keep it as it is, but I’m definitely open to discussing this approach further in our next meeting!
